apis/_codegen/fetch-apis.go

```diff
@@ -1,99 +1,235 @@
 	if err != nil {
-		fmt.Println("Failed to fetch api manifest, error: ", err)
-		os.Exit(1)
+		log.Fatalln("error: failed to fetch api manifest: ", err)
 	}
 	// Parse API manifest
 	var manifest map[string]string
 	if err = json.Unmarshal(res.Body, &manifest); err != nil {
-		fmt.Println("Failed to parse api manifest, error: ", err)
-		os.Exit(1)
+		log.Fatalln("error: failed to parse api manifest: ", err)
 	}
 
+	fmt.Println("Fetching Services:")
 	services := make(map[string]definitions.Service)
 	for name, referenceURL := range manifest {
-		fmt.Println(" - Fetching", name)
-		// Fetch reference
-		res, err2 := g.Get(referenceURL).Send()
-		if err2 != nil {
-			fmt.Println("Failed to fetch API: ", name, " error: ", err2)
-			os.Exit(1)
-		}
-		// Parse reference
-		var s definitions.Service
-		if err2 := json.Unmarshal(res.Body, &s); err2 != nil {
-			fmt.Println("Failed parse API: ", name, " error: ", err2)
-			os.Exit(1)
-		}
-		services[name] = s
+		wg.Add(1)
+		go func(n string, u string) {
+			s := fetchService(g, n, u)
+
+			mutex.Lock()
+			services[n] = s
+			mutex.Unlock()
+			wg.Done()
+		}(name, referenceURL)
 	}
+	wg.Wait()
+
+	gen.Print("var services = ")
+	gen.PrintComposite(services, true)
+	gen.Print("\n")
 
 	// Fetch all schemas
 	fmt.Println("Fetching Schemas:")
-	schemas := make(map[string]string)
-	fetchSchema := func(url string) {
-		if _, ok := schemas[url]; ok {
+	schemas := make(map[string]string, 0)
+	urls := make(map[string]bool, 0)
+
+	addSchema := func(url string) {
+		if url == "" || urls[url] {
 			return
 		}
-		fmt.Println(" - ", url)
-		res, err2 := g.Get(url).Send()
-		if err2 != nil {
-			fmt.Println("Failed to fetch: ", url, " error: ", err)
-			os.Exit(1)
-		}
-		// Test that we can parse the JSON schema (otherwise it's invalid)
-		var i interface{}
-		if json.Unmarshal(res.Body, &i) != nil {
-			fmt.Println("Failed to parse: ", url, " error: ", err)
-			os.Exit(1)
-		}
-		schemas[url] = string(res.Body)
+
+		urls[url] = true
+		wg.Add(1)
+		go func() {
+			s := fetchSchema(g, url)
+
+			mutex.Lock()
+			schemas[url] = s
+			mutex.Unlock()
+			wg.Done()
+		}()
 	}
 	for _, s := range services {
 		for _, e := range s.Entries {
-			if e.Input != "" {
-				fetchSchema(e.Input)
-			}
-			if e.Output != "" {
-				fetchSchema(e.Output)
-			}
+			addSchema(e.Input)
+			addSchema(e.Output)
 		}
 	}
+	wg.Wait()
 
-	code := "package apis\n"
-	code += "\n"
-	code += "import \"github.com/taskcluster/taskcluster-cli/apis/definitions\"\n"
-	code += "\n"
-	code += "var services = " + fmt.Sprintf("%#v", services)
-	code += "\n"
-	code += "var schemas = " + fmt.Sprintf("%#v", schemas)
-	code += "\n"
+	gen.Print("var schemas = ")
+	gen.PrintComposite(schemas, true)
+	gen.Print("\n")
 
-	source, err := format.Source([]byte(code))
+	// Format the output.
+	source, err := gen.Format()
 	if err != nil {
-		fmt.Println("go fmt, code generation failed, error: ", err)
-		os.Exit(1)
+		log.Println(gen.String())
```

Comment:
It does. Under normal circumstances that line should not be executed because we've ensured that the output we generate is valid go code. However it is useful for debugging the code when we make changes. I'll take it out.